README.md
# fib-retracement

A tool to easily calculate the fibonacci retracement levels based on both extremes (0 and 1).

## Status

[![npm version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]

## Installation

Install the package via `yarn`:

```sh
❯ yarn add fib-retracement
```

or via `npm`:

```sh
❯ npm install fib-retracement --save
```

## Usage

A single method is exported called `getFibRetracement`. For convenience `levels` are also exported in case you need them.

```js
const { getFibRetracement, levels } = require('fib-retracement');

/**
 * levels = [0, 0.236, 0.382, 0.5, 0.618, 0.786, 1]
 */

// Run with named parameter notation where 1 is the `start` and 0 is the `end`

const fib = getFibRetracement({ 0: 10, 1: 0 });

// Or with `start`, `end` positional notation

const fib = getFibRetracement(0, 10);

/**
 * fib = {
 *   1: 0
 *   0.786: 2.1399999999999997,
 *   0.618: 3.8200000000000003,
 *   0.5: 5,
 *   0.382: 6.18,
 *   0.236: 7.640000000000001,
 *   0: 10,
 * }
*/
```

__NOTE__: The method supports fibonacci retracement calculations as long as both extremes are provided, i.e. `0` and `1`.

## Tests

```
❯ yarn test
```

## License

MIT

[npm-image]: https://img.shields.io/npm/v/fib-retracement.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/fib-retracement
[travis-image]: https://img.shields.io/travis/madeiras/fib-retracement.svg?style=flat-square
[travis-url]: https://travis-ci.org/madeiras/fib-retracement


CHANGELOG.md
# Changelog

## [v0.2.1](https://github.com/madeiras/fib-retracement/releases/tag/v0.2.1) (2023-02-14)

## [v0.1.1](https://github.com/madeiras/fib-retracement/releases/tag/v0.1.1) (2023-02-01)

## [v0.1.0](https://github.com/madeiras/fib-retracement/releases/tag/v0.1.0) (2023-02-01)

## [v0.0.4](https://github.com/madeiras/fib-retracement/releases/tag/v0.0.4) (2021-05-27)

## [v0.0.3](https://github.com/madeiras/fib-retracement/releases/tag/v0.0.3) (2021-03-16)

## [v0.0.2](https://github.com/madeiras/fib-retracement/releases/tag/v0.0.2) (2020-05-02)

## [v0.0.1](https://github.com/madeiras/fib-retracement/releases/tag/v0.0.1) (2020-05-01)


index.spec.ts
import * as fibRetracement from './index';

/**
 * Test `fib-retracement`.
 */

describe('fib-retracement', () => {
  describe('getFibRetracement()', () => {
    ['foo', [], {}, 1, null, undefined].forEach(value => {
      it('should throw an error if `levels` is not a valid json object', () => {
        try {
          fibRetracement.getFibRetracement({ levels: value as unknown as Record<number| string, number | string> });

          fail();
        } catch (e) {
          expect(e.message).toBe('Unable to compute fib trace with the referenced `levels`');
        }
      });
    });

    ['foo', {}, NaN, undefined].forEach(value => {
      it('should return an empty object if the `Number` casted `0` is NaN', () => {
        expect(fibRetracement.getFibRetracement({ levels: { 0: value as unknown as number, 1: 1 } })).toEqual({});
      });
    });

    ['foo', {}, NaN, undefined].forEach(value => {
      it('should return an empty object if the `Number` casted `1` is NaN', () => {
        expect(fibRetracement.getFibRetracement({ levels: { 0: 0, 1: value as unknown as number } })).toEqual({});
      });
    });

    it('should convert the values to `Number`', () => {
      expect(fibRetracement.getFibRetracement({ levels: { 0: '4.74', 1: '4.59' } })).toEqual({
        0: 4.74,
        0.236: 4.7046,
        0.382: 4.6827,
        0.5: 4.665,
        0.618: 4.6473,
        0.786: 4.6221,
        1: 4.59
      });
    });

    it('should return the correct fib levels for downward fibs', () => {
      expect(fibRetracement.getFibRetracement({ levels: { 0: -61.43, 1: 70.1 } })).toEqual({
        0: -61.43000000000001,
        0.236: -30.388920000000013,
        0.382: -11.185540000000003,
        0.5: 4.334999999999994,
        0.618: 19.85553999999999,
        0.786: 41.95258,
        1: 70.1
      });
    });

    it('should return the correct fib keys for upward fibs', () => {
      expect(fibRetracement.getFibRetracement({ levels: { 0: 6427.56, 1: 6280.0 } })).toEqual({
        0: 6427.56,
        0.236: 6392.73584,
        0.382: 6371.19208,
        0.5: 6353.780000000001,
        0.618: 6336.367920000001,
        0.786: 6311.57784,
        1: 6280
      });
    });

    it('should accept call with positional parameters', () => {
      expect(fibRetracement.getFibRetracement('70.1', '-61.43')).toEqual({
        0: -61.43000000000001,
        0.236: -30.388920000000013,
        0.382: -11.185540000000003,
        0.5: 4.334999999999994,
        0.618: 19.85553999999999,
        0.786: 41.95258,
        1: 70.1
      });
    });
  });
});


index.test.js
/* eslint-disable sort-keys */
'use strict';

const fibRetracement = require('./index');

/**
 * Test `fib-retracement`.
 */

describe('fib-retracement', () => {
  describe('getFibRetracement()', () => {
    it('should throw an error if `levels` is not defined', () => {
      try {
        fibRetracement.getFibRetracement();

        fail();
      } catch (e) {
        expect(e.message).toBe('Unable to compute fib trace with the referenced `levels`');
      }
    });

    ['foo', [], {}, 1, null, undefined].forEach(value => {
      it('should throw an error if `levels` is not a valid json object', () => {
        try {
          fibRetracement.getFibRetracement({ levels: value });

          fail();
        } catch (e) {
          expect(e.message).toBe('Unable to compute fib trace with the referenced `levels`');
        }
      });
    });

    ['foo', {}, NaN, undefined].forEach(value => {
      it('should return an empty object if the `Number` casted `0` is NaN', () => {
        expect(fibRetracement.getFibRetracement({ levels: { 0: value, 1: 1 } })).toEqual({});
      });
    });

    ['foo', {}, NaN, undefined].forEach(value => {
      it('should return an empty object if the `Number` casted `1` is NaN', () => {
        expect(fibRetracement.getFibRetracement({ levels: { 0: 0, 1: value } })).toEqual({});
      });
    });

    it('should convert the values to `Number`', () => {
      expect(fibRetracement.getFibRetracement({ levels: { 0: '4.74', 1: '4.59' } })).toEqual({
        0: 4.74,
        0.236: 4.7046,
        0.382: 4.6827,
        0.5: 4.665,
        0.618: 4.6473,
        0.786: 4.6221,
        1: 4.59
      });
    });

    it('should return the correct fib levels for downward fibs', () => {
      expect(fibRetracement.getFibRetracement({ levels: { 0: -61.43, 1: 70.1 } })).toEqual({
        0: -61.43000000000001,
        0.236: -30.388920000000013,
        0.382: -11.185540000000003,
        0.5: 4.334999999999994,
        0.618: 19.85553999999999,
        0.786: 41.95258,
        1: 70.1
      });
    });

    it('should return the correct fib keys for upward fibs', () => {
      expect(fibRetracement.getFibRetracement({ levels: { 0: 6427.56, 1: 6280.0 } })).toEqual({
        0: 6427.56,
        0.236: 6392.73584,
        0.382: 6371.19208,
        0.5: 6353.780000000001,
        0.618: 6336.367920000001,
        0.786: 6311.57784,
        1: 6280
      });
    });
  });
});


index.js
'use strict';

/**
 * Exports.
 */

module.exports = {
  /**
   * Get the fib retracement based on `0` and `1` fib retracement levels.
   */

  getFibRetracement({ levels } = {}) {
    if (!levels || typeof levels !== 'object' || Array.isArray(levels) || Object.keys(levels).length === 0) {
      throw new Error('Unable to compute fib trace with the referenced `levels`');
    }

    const zero = Number(levels[0]);
    const one = Number(levels[1]);

    if (isNaN(zero) || isNaN(one)) {
      return {};
    }

    const distance = Math.abs(one - zero);
    const operation = one > zero ? (first, second) => first - second : (first, second) => first + second;

    return Object.fromEntries(
      module.exports.levels.map(level => {
        const difference = (1 - level) * distance;
        const value = operation(Number(levels[1]), difference);

        return [level, value];
      })
    );
  },

  /**
   * Levels.
   */

  levels: [0, 0.236, 0.382, 0.5, 0.618, 0.786, 1]
};


index.ts
/**
 * Calculate levels.
 */

function calculate({ levels }: CalculateInterface) {
  if (!levels || typeof levels !== 'object' || Array.isArray(levels) || Object.keys(levels).length === 0) {
    throw new Error('Unable to compute fib trace with the referenced `levels`');
  }

  const one = Number(levels[1]);
  const zero = Number(levels[0]);

  if (isNaN(one) || isNaN(zero)) {
    return {};
  }

  const distance = Math.abs(one - zero);
  const operation = one > zero ? 
    (first: number, second: number) => first - second : 
    (first: number, second: number) => first + second;

  return Object.fromEntries(
    module.exports.levels.map(level => {
      const difference = (1 - level) * distance;
      const value = operation(Number(levels[1]), difference);

      return [level, value];
    })
  );
}

interface CalculateInterface {
  levels: Record<number | string, number | string>
} 

/**
 * Get the fib retracement based on `0` and `1` fib retracement levels.
 */

export function getFibRetracement(start: number | string, end: number | string);
export function getFibRetracement({ levels }: CalculateInterface);
export function getFibRetracement(...values: unknown[]) {
  const payload = values[0] && Object.prototype.hasOwnProperty.call(values[0], 'levels') ? 
    values[0] : 
    { levels: { 0: values[1], 1: values[0] } }

  return calculate(payload as CalculateInterface);
}

/**
 * Levels.
 */

export const levels = [0, 0.236, 0.382, 0.5, 0.618, 0.786, 1];